#!/usr/bin/env bash

####Parameterized SLURM Script ####
#SBATCH --job-name="dysgu"
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --cpus-per-task=8
#SBATCH --mem=20G
#SBATCH --partition=defq
#SBATCH --time=0-00:40:00
#SBATCH --error="log_dysgu_jid%J.out"
#SBATCH --output="log_dysgu_jid%J.out"
#SBATCH --mail-type=FAIL,TIME_LIMIT,TIME_LIMIT_90,TIME_LIMIT_80,TIME_LIMIT_50
#SBATCH --mail-user=`whoami`@tgen.org

##@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
## HELP of the TOOL dysgu
# github: https://github.com/kcleal/dysgu
#singularity exec -B /scratch/,/home,/labs /home/tgenref/containers/grandcanyon/structural_variant_calling/dysgu_1.5.0--py310h770aed0_1.sif dysgu --help
#Usage: dysgu [OPTIONS] COMMAND [ARGS]...
#
#  Dysgu-SV is a set of tools calling structural variants from bam/cram files
#
#Options:
#  --version  Show the version and exit.
#  --help     Show this message and exit.
#
#Commands:
#  call    Call structural variants from bam alignment file/stdin
#  fetch   Filters input bam/cram for read-pairs that are discordant or...
#  filter  Filter a vcf generated by dysgu.
#  merge   Merge vcf/csv variant files
#  run     Run the dysgu pipeline.
#  test    Run dysgu tests
##@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

# INPUT TESTED:
# /home/tgenref/homo_sapiens/t2t_chm13/chm13_v2/genome_reference/chm13v2.0_maskedY_rCRS.fa
# COLO829BL_chm13.bam --> WORKED with fetch, call and filter tasks
# /scratch/jkeats/revio/COLO829BL_ATCCJJK_p1_CL_C2_mm_rg_sort_chm13.cram --> pacbio --> WORKED
# COLO829BL_ATCCJJK_p1_CL_C2_pbmmLike_rg_sort.cram --> pacbio --> WORKED
# KMS11_JCRBsus_p19_CL_Whole_T1_OGL14_mm_rg_sort_chm13.cram --> ONT --> worked


function usage(){
  echo "
USAGE: minimum information needed: Ref_genome and Bam file with pacbio data;
$0
1__REF_GENOME  (REQUIRED )
2__BAM  (REQUIRED)
3__OUTFILE (no default seen in help need test to see if there is any default
4__DIR_WORK (default: current directory)
5__TYPE_OF_READ [pe|pacbio|nanopore]
6__CPUS_for_Fetch (no need more than 6 based on first tests)

"
echo ''
}

function checkDir(){
        local D=$1
        if [[ ! -e ${D} ]] ; then echo -e "DIR NOT FOUND << ${D} >>; Aborting!" ; exit 1 ; fi ;
}
function checkFile(){
        local F=$1
        if [[ ! -e ${F} ]] ; then echo -e "FILE NOT FOUND << ${F} >>; Aborting!" ; exit 1 ; fi ;
}

function check_ev(){
  EV=$1 ; MSG=$2
  if [ ${EV} -ne 0 ] ; then echo -e "ERROR with: ${MSG}" ; exit 1 ; else echo "${MSG}: OK!" ; fi
}


module load singularity/3.8.4
MOUNTS="/scratch/,/home,/labs"
SIF_IMAGE_DYSGU="/home/tgenref/containers/grandcanyon/structural_variant_calling/dysgu_1.5.0--py310h770aed0_1.sif"
SIF_BCFTOOLS="/home/tgenref/containers/grandcanyon/common_tools/bcftools_1.17--h3cc50cf_1.sif"

ADD_OPTIONS=" "
## manage options based on given inputs

## REQUIRED ARGUMENTS
if [[ ${BED_FILE_TO_EXCLUDE} != "" ]] ;
then
	checkFile  ${BED_FILE_TO_EXCLUDE}
	ADD_OPTIONS=" ${ADD_OPTIONS[@]} --exclude ${BED_FILE_TO_EXCLUDE}" ;
fi

if [[ ${BED_FILE_TO_SEARCH} != "" ]] ;
then
	checkFile ${BED_FILE_TO_SEARCH}
	ADD_OPTIONS=" ${ADD_OPTIONS[@]} --bam ${BED_FILE_TO_SEARCH}" ;
fi



EXPECTED_NUMBER_ARGS=6
if [[ $# -ne ${EXPECTED_NUMBER_ARGS} ]] ; then echo -e "Expected ${EXPECTED_NUMBER_ARGS} args found $# ; Aborting;" ; usage ; exit 1 ; fi


# POSITIONAL ARGUMENTS ARE: WORKING_DIRECTORY BAM
## dysgu fetch [OPTIONS] WORKING_DIRECTORY BAM
REF_GENOME=$1
BAM=$2
BAM_DISCORDANT_READS_OUTFILE=$3
DIR_WORK=$4
TYPE_READS_PLATFORM="$5"
CPUS=$6


POSTFIX="_tempf_"  ; ## NO default values; must be Text
COMPRESSION=wb0  ; ## default value wb0
CLIP_LENGTH=15  ; ## default value is 15
MIN_MAP_QUAL=1  ; ## default value is 1
MIN_SIZE=30  ; ## default value is 30
MAX_COV=200  ;   ## default value is 200
PROCS=16   ; ## default value is 1
#TYPE_READS_PLATFORM=pacbio  ; ## default value is pe; possibe values are pe, pacbio or nanopore
if [[ "$CPUS" != "" && "$CPUS" =~ ^[0-9]*$ ]] ; then PROCS=$CPUS ; fi

checkFile ${BAM} ${REF_GENOME}

## check if outfile has .bam extension; must be a bam extension;
extension="${BAM_DISCORDANT_READS_OUTFILE##*.}"
if [[ ${extension} != "bam" ]] ; then echo -e "ERROR: the filename of the output MUST be '.bam'. Please modify your input ; Aborting." ; exit 1 ; fi


mycmd="singularity exec -B ${MOUNTS} ${SIF_IMAGE_DYSGU} dysgu fetch \
--reference '${REF_GENOME}' \
--output '${BAM_DISCORDANT_READS_OUTFILE}' \
--compression ${COMPRESSION} \
--write_all \
--clip-length ${CLIP_LENGTH} \
--mq ${MIN_MAP_QUAL} \
--min-size ${MIN_SIZE} \
--max-cov ${MAX_COV} \
--procs ${PROCS} \
--overwrite \
--pfix ${POSTFIX} \
--pl '${TYPE_READS_PLATFORM}' \
${ADD_OPTIONS[@]} '${DIR_WORK}' ${BAM}
"

echo -e "dysgu fetch ... "
echo -e "${mycmd}"

eval ${mycmd}

check_ev $? "dysgu FETCH ... "
echo -e "dysgu fetch ... DONE "


#singularity exec -B /scratch/,/home,/labs /home/tgenref/containers/grandcanyon/structural_variant_calling/dysgu_1.5.0--py310h770aed0_1.sif dysgu fetch --help
#Usage: dysgu fetch [OPTIONS] WORKING_DIRECTORY BAM
#
#  Filters input bam/cram for read-pairs that are discordant or have a soft-
#  clip of length > '--clip-length', saves bam file in WORKING_DIRECTORY
#
#Options:
#  --reference PATH           Reference file for opening cram files
#  --pfix TEXT                Post-fix to add to temp alignment files
#  -o, --output TEXT          Output reads, discordant, supplementary and soft-
#                             clipped reads to file.
#  --compression TEXT         Set output bam compression level. Default is
#                             uncompressed  [default: wb0]
#  -a, --write_all            Write all alignments from SV-read template to
#                             temp file
#  --clip-length INTEGER      Minimum soft-clip length, >= threshold are kept.
#                             Set to -1 to ignore  [default: 15]
#  --mq INTEGER               Minimum map quality < threshold are discarded
#                             [default: 1]
#  --min-size INTEGER         Minimum size of SV to report  [default: 30]
#  --max-cov FLOAT            Genomic regions with coverage > max-cov are
#                             discarded. Set to -1 to ignore.  [default: 200]
#  -p, --procs INTEGER RANGE  Compression threads to use for writing bam
#                             [default: 1; 1<=x<=40]
#  --search PATH              .bed file, limit search to regions
#  --exclude PATH             .bed file, do not search/call SVs within regions.
#                             Takes precedence over --search
#  -x, --overwrite            Overwrite temp files
#  --pl [pe|pacbio|nanopore]  Type of input reads  [default: pe]
#  --help                     Show this message and exit.



#@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
# POSITIONAL ARGUMENTS ARE: REFERENCE_GENOME WORKING_DIRECTORY BAM_FILE_With_DISCORDANT_READS
## dysgu call [OPTIONS] REFERENCE_GENOME WORKING_DIRECTORY BAM_FILE_With_DISCORDANT_READS
#@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
SV_ALIGNS="${DISCORDANT_READS_OUTFILE}"

OUT_FORMAT='vcf'  ; ## default is vcf
POSTFIX="_tempf_"  ; ## NO default values; must be Text
COMPRESSION=wb0  ; ## default value wb0
MODE=${TYPE_READS_PLATFORM}  ; ## default value is pe [possible values are pe|pacbio|nanopore]
CLIP_LENGTH=15  ; ## default value is 15
MIN_MAP_QUAL=1  ; ## default value is 1
MIN_SIZE=30  ; ## default value is 30
MAX_COV=200  ;   ## default value is 200
PROCS=4   ; ## default value is 1
TYPE_READS_PLATFORM=${TYPE_READS_PLATFORM}  ; ## default value is pe; possibe values are pe, pacbio or nanopore
if [[ ${TYPE_READS_PLATFORM} != "pe" ]] ; then ADD_OPTIONS=" --paired False " ; fi

OPTIONS_WITH_DEFAULT_VALUES=" --sites-prob 0.6  --sites-pass-only True --parse-probs False --all-sites False  \
--min-support 3 --max-tlen 1000  --dist-norm 100.00 --spd 0.30 --trust-ins-len True  --length-extend 15  --divergence 0.2 \
--regions-only False --regions-mm-only False --buffer-size 0 --merge-within True  --drop-gaps True --merge-dist 500 --contigs True  --verbosity 1 \
--diploid True --remap True --keep-small --thresholds 0.45,0.45,0.45,0.45,0.45 "

##
## options NOT DEFINED in our command line because they are optional and related to specific conditions/methods/analysis
## --template-size : not used for pacbio or nanopore
## --regions
## --metrics
## --no-gt
##

mycmd="singularity exec -B ${MOUNTS}  ${SIF_IMAGE_DYSGU} dysgu call \
--ibam '${BAM}' \
--svs-out '${BAM_DISCORDANT_READS_OUTFILE}.sv.vcf' \
--out-format ${OUT_FORMAT} \
--pfix ${POSTFIX} \
--mode ${MODE} \
--pl ${TYPE_READS_PLATFORM} \
--clip-length ${CLIP_LENGTH} \
--mq ${MIN_MAP_QUAL} \
--min-size ${MIN_SIZE} \
--max-cov ${MAX_COV} \
--procs ${PROCS} \
--overwrite \
--pl '${TYPE_READS_PLATFORM}' \
${OPTIONS_WITH_DEFAULT_VALUES} \
${ADD_OPTIONS[@]} \
'${REF_GENOME}' \
'${DIR_WORK}' \
'${BAM_DISCORDANT_READS_OUTFILE}'
"

echo -e "dysgu call ... "
echo -e "${mycmd}"

eval ${mycmd}

check_ev $? "dysgu CALL ... "
echo -e "dysgu call ... DONE "



#singularity exec -B /scratch/,/home,/labs /home/tgenref/containers/grandcanyon/structural_variant_calling/dysgu_1.5.0--py310h770aed0_1.sif dysgu call --help
#Usage: dysgu call [OPTIONS] REFERENCE WORKING_DIRECTORY [SV_ALIGNS]
#
#  Call structural variants from bam alignment file/stdin
#
#Options:
#  -b, --ibam PATH                 Original input file usef with 'fetch'
#                                  command, used for calculating insert size
#                                  parameters
#  -o, --svs-out PATH              Output file [default: stdout]
#  -f, --out-format [csv|vcf]      Output format  [default: vcf]
#  --sites PATH                    A vcf file of known variant sites. Matching
#                                  output variants are labelled with 'PASS'
#                                  plus the ID from --sites
#  --sites-prob FLOAT RANGE        Prior probability that a matching variant in
#                                  --sites is true  [default: 0.6; 0<=x<=1]
#  --sites-pass-only [True|False]  Only add variants from sites that have PASS
#                                  [default: True]
#  --parse-probs [True|False]      Parse INFO:MeanPROB or FORMAT:PROB instead
#                                  of using --sites-p  [default: False]
#  --all-sites [True|False]        Output a genotype for all variants in
#                                  --sites (including homozygous reference 0/0)
#  --pfix TEXT                     Post-fix of temp alignment file (used when a
#                                  working-directory is provided instead of sv-
#                                  aligns)
#  --mode [pe|pacbio|nanopore]     Type of input reads. Multiple options are
#                                  set, overrides other options
#                                  pacbio/nanopore: --mq 20 --paired False
#                                  --min-support 3 --max-cov 150  [default: pe]
#  --pl [pe|pacbio|nanopore]       Type of input reads  [default: pe]
#  --clip-length INTEGER           Minimum soft-clip length, >= threshold are
#                                  kept. Set to -1 to ignore  [default: 15]
#  --max-cov TEXT                  Regions with > max-cov that do no overlap
#                                  'include' are discarded. Use 'auto' to
#                                  estimate a value from the alignment index
#                                  file [default: 200]. Regions with > max-cov
#                                  that do no overlap 'include' are discarded.
#                                  Set to -1 to ignore.
#  --max-tlen INTEGER              Maximum template length to consider when
#                                  calculating paired-end template size
#                                  [default: 1000]
#  --min-support INTEGER           Minimum number of reads per SV  [default: 3]
#  --min-size INTEGER              Minimum size of SV to report  [default: 30]
#  --mq INTEGER                    Minimum map quality < threshold are
#                                  discarded  [default: 1]
#  --dist-norm FLOAT               Distance normalizer  [default: 100]
#  --spd FLOAT                     Span position distance  [default: 0.3]
#  --trust-ins-len TEXT            Trust insertion length from cigar, for high
#                                  error rate reads use False  [default: True]
#  --length-extend INTEGER         Extend SV length if any nearby gaps found
#                                  with length >= length-extend. Ignored for
#                                  paired-end reads  [default: 15]
#  --divergence TEXT               Threshold used for ignoring divergent end-
#                                  regions of alignments. Ignored for paired-
#                                  end reads  [default: 0.02]
#  -I, --template-size TEXT        Manually set insert size, insert stdev,
#                                  read_length as 'INT,INT,INT'
#  --regions PATH                  bed file of target regions, used for
#                                  labelling events
#  --regions-only [True|False]     If --regions is provided, call only events
#                                  within target regions  [default: False]
#  --regions-mm-only [True|False]  If --regions is provided, only use minimizer
#                                  clustering within --regions. Useful for high
#                                  coverage targeted sequencing  [default:
#                                  False]
#  -p, --procs INTEGER RANGE       Processors to use  [default: 1; 1<=x<=40]
#  --buffer-size INTEGER           Number of alignments to buffer  [default: 0]
#  --merge-within [True|False]     Try and merge similar events, recommended
#                                  for most situations  [default: True]
#  --drop-gaps [True|False]        Drop SVs near gaps +/- 250 bp of Ns in
#                                  reference  [default: True]
#  --merge-dist INTEGER            Attempt merging of SVs below this distance
#                                  threshold, default is (insert-median +
#                                  5*insert_std) for pairedreads, or 500 bp for
#                                  single-end reads
#  --paired [True|False]           Paired-end reads or single  [default: True]
#  --contigs [True|False]          Generate consensus contigs for each side of
#                                  break and use sequence-based metrics in
#                                  model scoring  [default: True]
#  -v, --verbosity [0|1|2]         0 = no contigs in output, 1 = output contigs
#                                  for variants without ALT sequence called, 2
#                                  = output all contigs  [default: 1]
#  --diploid [True|False]          Use diploid model for scoring variants. Use
#                                  'False' for non-diploid or poly clonal
#                                  samples  [default: True]
#  --remap TEXT                    Try and remap anomalous contigs to find
#                                  additional small SVs  [default: True]
#  --metrics                       Output additional metrics for each SV
#  --no-gt                         Skip adding genotype to SVs
#  --keep-small                    Keep SVs < min-size found during re-mapping
#  --low-mem                       Use less memory but more temp disk space
#  -x, --overwrite                 Overwrite temp files
#  -c, --clean                     Remove temp files when finished
#  --thresholds TEXT               Probability threshold to label as PASS for
#                                  'DEL,INS,INV,DUP,TRA'  [default:
#                                  0.45,0.45,0.45,0.45,0.45]
#  --help                          Show this message and exit.

MIN_PROB=0.2 ; ## default value is 0.2 for LongReads
SUPPORT_FRACTION=0.15 ; ## default value is 0.15 for LongReads
PASS_PROB=0.3 ; ## default value is 0.3 for LongReads

echo -e "PWD == $PWD"
mycmd="singularity exec -B ${MOUNTS} ${SIF_IMAGE_DYSGU} dysgu filter \
--min-prob ${MIN_PROB} \
--support-fraction ${SUPPORT_FRACTION} \
--pass-prob ${PASS_PROB} \
'${BAM_DISCORDANT_READS_OUTFILE}.sv.vcf' > '${BAM_DISCORDANT_READS_OUTFILE}.sv.filter.vcf' "

echo -e "dysgu filter for germline/tumor_only without normal or pon ... "
echo ${mycmd}
eval ${mycmd}
check_ev $? "dysgu FILTER ..."
echo -e "dysgu filter for germline/tumor_only without normal or pon ... DONE "





## compressed filtered VCF
## NOTE: with bcftools version 1.18, you only need one line ... there is an option to write index while converting ...
echo -e "bcftools conversion vcf to vcf.gz ..."
echo -e "PWD == $PWD"
singularity exec -B ${MOUNTS} ${SIF_BCFTOOLS} bcftools view --threads ${CPUS} -O z -o ${BAM_DISCORDANT_READS_OUTFILE}.sv.filter.vcf.gz ${BAM_DISCORDANT_READS_OUTFILE}.sv.filter.vcf

check_ev $? "bcftools conversion vcf->vcf.gz"

echo -e "bcftools indexing ..."

singularity exec -B ${MOUNTS}  ${SIF_BCFTOOLS} bcftools index --threads ${CPUS} --tbi ${BAM_DISCORDANT_READS_OUTFILE}.sv.filter.vcf.gz
check_ev $? "bcftools index tbi"

singularity exec  -B ${MOUNTS} ${SIF_BCFTOOLS} bcftools index --threads ${CPUS}  ${BAM_DISCORDANT_READS_OUTFILE}.sv.filter.vcf.gz
check_ev $? "bcftools index csi"


#singularity exec -B /scratch/,/home,/labs /home/tgenref/containers/grandcanyon/structural_variant_calling/dysgu_1.5.0--py310h770aed0_1.sif dysgu filter --help
#Usage: dysgu filter [OPTIONS] INPUT_VCF [NORMAL_BAMS]...
#
#  Filter a vcf generated by dysgu. Unique SVs can be found in the input_vcf by
#  supplying a --normal-vcf (single or multi-sample), and normal bam files.
#  Bam/vcf samples with the same name as the input_vcf will be ignored
#
#Options:
#  --reference PATH              Reference for cram input files
#  -o, --svs-out PATH            Output file, [default: stdout]
#  -n, --normal-vcf PATH         Vcf file for normal sample, or panel of
#                                normals. The SM tag of input bams is used to
#                                ignore the input_vcf for multi-sample vcfs
#  -p, --procs INTEGER RANGE     Reading threads for normal_bams  [default: 1;
#                                1<=x<=40]
#  -f, --support-fraction FLOAT  Minimum threshold support fraction / coverage
#                                (SU/COV)  [default: 0.1]
#  --target-sample TEXT          If input_vcf if multi-sample, use target-
#                                sample as input
#  --keep-all                    All SVs classified as normal will be kept in
#                                the output, labelled as filter=normal
#  --ignore-read-groups          Ignore ReadGroup RG tags when parsing sample
#                                names. Filenames will be used instead
#  --min-prob FLOAT              Remove SVs with PROB value < min-prob
#                                [default: 0.1]
#  --pass-prob FLOAT             Re-label SVs as PASS if PROB value >= pass-
#                                prob  [default: 1.0]
#  --interval-size INTEGER       Interval size for searching normal-vcf/normal-
#                                bams  [default: 1000]
#  --random-bam-sample INTEGER   Choose N random normal-bams to search. Use -1
#                                to ignore  [default: -1]
#  --help                        Show this message and exit.






# Unused but good to know that it exists
#@@@@@@@@@@@@@@@@#
####  UNSUSED ####
#@@@@@@@@@@@@@@@@#
#singularity exec -B /scratch/,/home,/labs /home/tgenref/containers/grandcanyon/structural_variant_calling/dysgu_1.5.0--py310h770aed0_1.sif dysgu merge --help
#Usage: dysgu merge [OPTIONS] INPUT_FILES...
#
#  Merge vcf/csv variant files
#
#Options:
#  -o, --svs-out PATH              Output file, [default: stdout]
#  -f, --out-format [csv|vcf]      Output format  [default: vcf]
#  --collapse-nearby [True|False]  Merges more aggressively by collapsing
#                                  nearby SV  [default: True]
#  --merge-across [True|False]     Merge records across input samples
#                                  [default: True]
#  --merge-within [True|False]     Perform additional merge within input
#                                  samples, prior to --merge-across  [default:
#                                  False]
#  --merge-dist INTEGER            Distance threshold for merging  [default:
#                                  500]
#  --separate [True|False]         Keep merged tables separate, adds --post-fix
#                                  to file names, csv format only  [default:
#                                  False]
#  --post-fix TEXT                 Adds --post-fix to file names, only if
#                                  --separate is True  [default: dysgu]
#  --add-kind [True|False]         Add region-overlap 'kind' to vcf output
#                                  [default: False]
#  -v, --verbosity [0|1|2]         0 = no contigs in output, 1 = output contigs
#                                  for variants without ALT sequence called, 2
#                                  = output all contigs  [default: 1]
#  --help                          Show this message and exit.



#@@@@@@@@@@@@@@@@#
####  UNSUSED ####
#@@@@@@@@@@@@@@@@#
## dysgu run; NOT IMPLEMENTED and USED so far because we want more control over the steps fetch and call and filter steps.

#singularity exec -B /scratch/,/home,/labs /home/tgenref/containers/grandcanyon/structural_variant_calling/dysgu_1.5.0--py310h770aed0_1.sif dysgu run --help
#Usage: dysgu run [OPTIONS] REFERENCE WORKING_DIRECTORY BAM
#
#  Run the dysgu pipeline. Important parameters are --mode, --diploid, --min-
#  support, --min-size, --max-cov
#
#Options:
#  --sites PATH                    A vcf file of known variant sites. All sites
#                                  will be genotyped in the output vcf
#  --sites-prob FLOAT RANGE        Prior probability that a matching variant in
#                                  --sites is true  [default: 0.6; 0<=x<=1]
#  --sites-pass-only [True|False]  Only add variants from sites that have PASS
#                                  [default: True]
#  --parse-probs [True|False]      Parse INFO:MeanPROB or FORMAT:PROB instead
#                                  of using --sites-p  [default: False]
#  --all-sites [True|False]        Output a genotype for all variants in
#                                  --sites (including homozygous reference 0/0)
#  --pfix TEXT                     Post-fix to add to temp alignment files
#  -o, --svs-out PATH              Output file, [default: stdout]
#  -f, --out-format [csv|vcf]      Output format  [default: vcf]
#  -a, --write_all                 Write all alignments from SV-read template
#                                  to temp file
#  --compression TEXT              Set temp file bam compression level. Default
#                                  is uncompressed  [default: wb0]
#  -p, --procs INTEGER RANGE       Number of cpu cores to use  [default: 1;
#                                  1<=x<=40]
#  --mode [pe|pacbio|nanopore]     Type of input reads. Multiple options are
#                                  set, overrides other options. pacbio: --mq
#                                  20 --paired False --min-support 3 --max-cov
#                                  150 --dist-norm 200 --trust-ins-len True.
#                                  nanopore: --mq 20 --paired False --min-
#                                  support 3 --max-cov 150 --dist-norm 900
#                                  --trust-ins-len False  [default: pe]
#  --pl [pe|pacbio|nanopore]       Type of input reads  [default: pe]
#  --clip-length INTEGER           Minimum soft-clip length, >= threshold are
#                                  kept. Set to -1 to ignore  [default: 15]
#  --max-cov TEXT                  Genomic regions with coverage > max-cov
#                                  discarded. Use 'auto' to estimate a value
#                                  from the alignment index file [default:
#                                  200]. Set to -1 to ignore
#  --max-tlen INTEGER              Maximum template length to consider when
#                                  calculating paired-end template size
#                                  [default: 1000]
#  --min-support INTEGER           Minimum number of reads per SV  [default: 3]
#  --min-size INTEGER              Minimum size of SV to report  [default: 30]
#  --mq INTEGER                    Minimum map quality < threshold are
#                                  discarded  [default: 1]
#  --dist-norm FLOAT               Distance normalizer  [default: 100]
#  --spd FLOAT                     Span position distance  [default: 0.3]
#  --trust-ins-len TEXT            Trust insertion length from cigar, for high
#                                  error rate reads use False  [default: True]
#  --length-extend INTEGER         Extend SV length if any nearby gaps found
#                                  with length >= length-extend. Ignored for
#                                  paired-end reads  [default: 15]
#  --divergence TEXT               Threshold used for ignoring divergent ends
#                                  of alignments. Ignored for paired-end reads.
#                                  Use 'auto' to try to infer for noisy reads
#                                  [default: 0.02]
#  -I, --template-size TEXT        Manually set insert size, insert stdev,
#                                  read_length as 'INT,INT,INT'
#  --search PATH                   .bed file, limit search to regions
#  --exclude PATH                  .bed file, do not search/call SVs within
#                                  regions. Takes precedence over --search
#  --regions PATH                  bed file of target regions, used for
#                                  labelling events
#  --regions-mm-only [True|False]  If --regions is provided, only use minimizer
#                                  clustering within --regions. Useful for high
#                                  coverage targeted sequencing  [default:
#                                  False]
#  --buffer-size INTEGER           Number of alignments to buffer  [default: 0]
#  --merge-within [True|False]     Try and merge similar events, recommended
#                                  for most situations  [default: True]
#  --drop-gaps [True|False]        Drop SVs near gaps +/- 250 bp of Ns in
#                                  reference  [default: True]
#  --merge-dist INTEGER            Attempt merging of SVs below this distance
#                                  threshold. Default for paired-end data is
#                                  (insert-median + 5*insert_std) for
#                                  pairedreads, or 500 bp for single-end reads
#  --paired [True|False]           Paired-end reads or single  [default: True]
#  --contigs [True|False]          Generate consensus contigs for each side of
#                                  break and use sequence-based metrics in
#                                  model scoring  [default: True]
#  -v, --verbosity [0|1|2]         0 = no contigs in output, 1 = output contigs
#                                  for variants without ALT sequence called, 2
#                                  = output all contigs  [default: 1]
#  --diploid [True|False]          Use diploid model for scoring variants. Use
#                                  'False' for non-diploid or poly clonal
#                                  samples  [default: True]
#  --remap TEXT                    Try and remap anomalous contigs to find
#                                  additional small SVs  [default: True]
#  --metrics                       Output additional metrics for each SV
#  --no-gt                         Skip adding genotype to SVs
#  --keep-small                    Keep SVs < min-size found during re-mapping
#  --low-mem                       Use less memory but more temp disk space
#  -x, --overwrite                 Overwrite temp files
#  -c, --clean                     Remove temp files when finished
#  --thresholds TEXT               Probability threshold to label as PASS for
#                                  'DEL,INS,INV,DUP,TRA'  [default:
#                                  0.45,0.45,0.45,0.45,0.45]
#  --help                          Show this message and exit.
